### PR TITLE
discovery: Make status updates non blocking

### PR DESCRIPTION
**Pull request by charlieegan3**
Fixes https://github.com/open-policy-agent/opa/issues/6343

In addition to what's changed here, we might want to consider making the status posts have a configurable timeout, but that'd be a change to config and something we might want to do in another PR.

### CODE REVIEW

File: plugins/status/plugin.go [hunk 1]
```diff
@@ -204,7 +205,9 @@ func New(parsedConfig *Config, manager *plugins.Manager) *Plugin {
 		decisionLogsCh: make(chan lstat.Status),
 		stop:           make(chan chan struct{}),
 		reconfig:       make(chan reconfigure),
-		pluginStatusCh: make(chan map[string]*plugins.Status),
+		// we use a buffered channel here to avoid blocking other plugins
+		// when updating statuses
+		pluginStatusCh: make(chan map[string]*plugins.Status, 3),
```
Comment (charlieegan3) [inline plugins/status/plugin.go:210]: This is the other part of the fix, https://github.com/open-policy-agent/opa/issues/6343#issuecomment-1780266088

Comment (ashutosh-narkar) [reply plugins/status/plugin.go:210]: Curious how we can up with the `3`?

Comment (charlieegan3) [reply plugins/status/plugin.go]: Actually, all we need to address the issue is 1, so I have used that instead in 908b57067100420fa2ae6f835bf360ccb41fe22d

File: plugins/discovery/discovery_test.go [hunk 2]
```diff
@@ -1557,24 +1546,60 @@ func TestStatusUpdates(t *testing.T) {
...
+		select {
+		case update := <-ts.updates:
+			if update.Discovery != nil {
+				matches := false
+				if update.Discovery.Code == nextExpectedDiscoveryUpdate.Code &&
+					update.Discovery.ActiveRevision == nextExpectedDiscoveryUpdate.Revision {
+					matches = true
+				}
+
+				if matches {
```
Comment (ashutosh-narkar) [inline plugins/discovery/discovery_test.go:1591]: Is it an error if there is no match?

File: plugins/status/plugin.go [hunk 3]
```diff
@@ -204,7 +205,9 @@ func New(parsedConfig *Config, manager *plugins.Manager) *Plugin {
 		decisionLogsCh: make(chan lstat.Status),
 		stop:           make(chan chan struct{}),
 		reconfig:       make(chan reconfigure),
-		pluginStatusCh: make(chan map[string]*plugins.Status),
+		// we use a buffered channel here to avoid blocking other plugins
+		// when updating statuses
+		pluginStatusCh: make(chan map[string]*plugins.Status, 1),
```
Comment (ashutosh-narkar) [inline plugins/status/plugin.go:210]: This would hold only one value, so wouldn't we have the same issue?

**Review by ashutosh-narkar — COMMENTED** [review]
@charlieegan3 the changes look fine. It would be helpful if we have test coverage for the scenario we're trying to fix here ie. multiple plugins updating their status and checking OPA startup.

**Comment by charlieegan3** [comment]
Thanks for the review Ash, I didn't managed to get a working test for this today but worked on it a bit. I'll try and get an update to you early next week.

File: plugins/discovery/discovery_test.go (hunk 2 above)
Comment (charlieegan3) [reply plugins/discovery/discovery_test.go:1592]: Not really, as the next status might be the expected one and then it's fine. This is just reading statuses until the expected one is found, before moving on to the next one.

**Comment by charlieegan3** [comment]
I've added a test now that exercises the fix. Without the fixes, the test fails due to the timeout.

File: plugins/discovery/discovery_test.go [hunk 4]
```diff
@@ -1557,24 +1547,171 @@ func TestStatusUpdates(t *testing.T) {
...
+		if err != nil {
+			t.Log(err)
+			return
+		}
+		booted <- true
+	}()
+
+	select {
+	case <-booted:
+		t.Log("OK, boot was not delayed")
```
Comment (ashutosh-narkar) [inline plugins/discovery/discovery_test.go:1712]: Another thing we can check here is the status of the plugins at least on the manager if not the status plugin endpoint. The later would need a custom status plugin to test I think. Anyways checking the plugin status should be useful.

**Review by ashutosh-narkar — COMMENTED** [review]
@charlieegan3 thanks for adding the test. One comment on the test and we can get this in.

**Review by ashutosh-narkar — APPROVED** [review]
LGTM. Please squash your commits and we can get this in. Thanks @charlieegan3!

**Comment by charlieegan3** [comment]
Thanks for the reviews Ash 🙏

**Pull request merged**